libminifi/src/utils/LineByLineInputOutputStreamCallback.cpp

```diff
@@ -0,0 +1,129 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "utils/LineByLineInputOutputStreamCallback.h"
+
+#include "utils/gsl.h"
+
+namespace org::apache::nifi::minifi::utils {
+
+LineByLineInputOutputStreamCallback::LineByLineInputOutputStreamCallback(uint64_t buffer_size, std::shared_ptr<core::logging::Logger> logger, CallbackType callback)
+  : buffer_(buffer_size),
+    logger_(std::move(logger)),
+    callback_(std::move(callback)) {
+}
+
+int64_t LineByLineInputOutputStreamCallback::process(const std::shared_ptr<io::BaseStream>& input, const std::shared_ptr<io::BaseStream>& output) {
+  gsl_Expects(input);
+  gsl_Expects(output);
+
+  input_size_ = input->size();
+
+  if (int64_t status = readLine(*input); status <= 0) {
+    return status;
+  }
+
+  bool is_first_line = true;
+  do {
+    if (int64_t status = readLine(*input); status < 0) {
+      return status;
+    }
+    std::string output_line = callback_(*current_line_, is_first_line, isLastLine());
+    output->write(reinterpret_cast<const uint8_t *>(output_line.data()), output_line.size());
+    is_first_line = false;
+  } while (!isLastLine());
+
+  return gsl::narrow<int64_t>(total_bytes_read_);
+}
+
+int64_t LineByLineInputOutputStreamCallback::readLine(io::InputStream& stream) {
+  std::string input_line;
+  const size_t current_read = readLine(stream, input_line);
+  if (io::isError(current_read)) {
+    return -1;
+  }
+  if (current_read == 0) {
+    current_line_ = next_line_;
+    next_line_ = std::nullopt;
+    return 0;
+  }
+  current_line_ = next_line_;
+  next_line_ = input_line;
+  total_bytes_read_ += current_read;
+  return gsl::narrow<int64_t>(current_read);
+}
+
+size_t LineByLineInputOutputStreamCallback::readLine(io::InputStream& stream, std::string& line) {
+  if (current_pos_ == end_pos_) {
+    const auto current_bytes_read = fillBuffer(stream);
+    if (io::isError(current_bytes_read) || current_bytes_read == 0) {
+      return current_bytes_read;
+    }
+  }
+
+  int64_t pos = current_pos_;
+  while (pos < end_pos_ && buffer_[pos] != '\n') {
+    ++pos;
+  }
+  if (pos < end_pos_ && buffer_[pos] == '\n') {
+    ++pos;
+    line = std::string{buffer_.begin() + current_pos_, buffer_.begin() + pos};
+    current_pos_ = pos;
+    return line.size();
+  }
+
+  std::string start_of_line{buffer_.begin() + current_pos_, buffer_.begin() + end_pos_};
+
+  const auto current_bytes_read = fillBuffer(stream);
+  if (io::isError(current_bytes_read)) {
+    return io::STREAM_ERROR;
+  } else if (current_bytes_read == 0) {
+    line = start_of_line;
+    return line.size();
+  }
+
+  pos = current_pos_;
+  while (pos < end_pos_ && buffer_[pos] != '\n') {
```

Comment:
unfortunately it seems nifi has a different concept of "line", it  uses the `LineDemarcator` which

1. considers all `\n` as end-of-line
2. considers all `\r` followed by a non-`\n` as end-of-line
3. considers a non-empty (!) tail after the last end-of-line as a line

(it's quite confusing) 

this was implemented in RouteText to be in-line with nifi, but of course we could decide to change them to use a different segmentation logic

Reply:
Just speculating, but that may have been done in NiFi to deal with the old MacOS line endings of bare `\r`.  Since these line endings are not used by any computer in use today, I think using `\n`, which covers Windows, Linux and MacOS X+, is fine.

Reply:
(3) is handled correctly (the `\n`-less last line is treated as a line, is processed and output without a `\n`), **except** in the edge case where the last line doesn't end in a `\n` **and** it straddles a buffer boundary, which causes an error (flow file routed to failure).  Thanks for pointing this out, I'll fix it.

Reply:
the bug in (3) is fixed by https://github.com/apache/nifi-minifi-cpp/pull/1170/commits/4df70bdec67d7f91182e10b01e6df646acf5e85a